tchannel-core/src/main/java/com/uber/tchannel/channels/ChannelManager.java

```diff
@@ -41,47 +39,102 @@
  * entry point for shutting down all active Channels.
  */
 public class ChannelManager {
+    private final Map<SocketAddress, Peer> peers = new Hashtable<>();
+    private String host = "0.0.0.0";
+    private int port = 0;
 
-    private final Map<InetSocketAddress, ChannelId> channelMap = new HashMap<>();
-    private final ChannelGroup channels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    public Connection findOrNew(SocketAddress address, Bootstrap bootstrap) throws InterruptedException {
+        Peer peer = peers.get(address);
+        if (peer == null) {
+            synchronized (peers) {
+                peer = new Peer(this, address);
+                peer.connect(bootstrap);
+                peers.put(address, peer);
+            }
+        }
 
-    public boolean add(Channel channel) {
-        this.channelMap.put((InetSocketAddress) channel.remoteAddress(), channel.id());
-        return this.channels.add(channel);
+        Connection conn = peer.getConnection(ConnectionState.IDENTIFIED);
+        return conn;
     }
 
-    public boolean remove(Channel channel) {
-        this.channelMap.remove(channel.remoteAddress());
-        return this.channels.remove(channel);
+    public Connection get(Channel channel) {
+        SocketAddress address = channel.remoteAddress();
+        Peer peer = peers.get(address);
+        if (peer != null) {
+            return peer.getConnection(channel.id());
+        }
+
+        return null;
     }
 
-    public Channel findOrNew(InetSocketAddress address, Bootstrap bootstrap) throws InterruptedException {
-        Channel channel = this.find(address);
+    public Channel add(ChannelHandlerContext ctx) {
+        Channel channel = ctx.channel();
+        System.out.println("Connection is active to " + channel.remoteAddress().toString());
 
-        if (channel == null) {
-            channel = this.newChannel(address, bootstrap);
+        SocketAddress address = channel.remoteAddress();
+        Peer peer = peers.get(address);
+        if (peer == null) {
+            synchronized (peers) {
+                peer = new Peer(this, address);
+                peers.put(address, peer);
+            }
         }
 
-        return channel;
+        return peer.handleActiveConnection(ctx, getDirection(ctx)).channel;
     }
 
-    public void close() throws InterruptedException {
-        this.channels.close().sync();
-        this.channels.clear();
-        this.channelMap.clear();
+    public static Connection.Direction getDirection(ChannelHandlerContext ctx) {
+        if (ctx.pipeline().names().contains("InitRequestHandler")) {
```

Comment:
The `InitRequestHandler` removes itself from the pipeline after the handshake so this logic may be incorrect.